src/alerter.rs

```diff
@@ -0,0 +1,389 @@
+use crate::WatchtowerConfig;
+use crate::pagerduty::PagerDutyClient;
+
+use anyhow::Result;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, Instant};
+use tokio::sync::mpsc::{self, UnboundedSender, UnboundedReceiver};
+use tokio::sync::Mutex;
+
+#[derive(Deserialize, Clone, PartialEq, Eq, Debug, Default)]
+pub enum AlertLevel {
+    #[default]
+    None,
+    Info,
+    Warn,
+    Error,
+}
+
+#[derive(Clone, Debug)]
+pub struct AlertParams {
+    name: String,
+    description: String,
+    alert_level: AlertLevel,
+}
+
+impl AlertParams {
+    pub fn new(name: String, description: String, alert_level: AlertLevel) -> Self {
+        AlertParams { name, description, alert_level }
+    }
+
+    #[cfg(test)]
+    pub fn is_name_equal(&self, name: &str) -> bool {
+        self.name == name
+    }
+
+    #[cfg(test)]
+    pub fn is_description_equal(&self, description: &str) -> bool {
+        self.description == description
+    }
+
+    #[cfg(test)]
+    pub fn is_level_equal(&self, alert_level: AlertLevel) -> bool {
+        self.alert_level == alert_level
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct WatchtowerAlerter{
+    alert_sender: UnboundedSender<AlertParams>,
+    alert_receiver: Arc<Mutex<UnboundedReceiver<AlertParams>>>,
+    alert_cache: Arc<Mutex<HashMap<String, Instant>>>,
+    alert_cache_expiry: Duration,
+    pagerduty_client: Option<PagerDutyClient>,
+    watchtower_system_name: String,
+    allowed_alerting_start_time: SystemTime,
+}
+
+impl WatchtowerAlerter{
+    pub fn new(config: &WatchtowerConfig, pagerduty_client: Option<PagerDutyClient>) -> Result<Self> {
+        let alert_cache = Arc::new(Mutex::new(HashMap::with_capacity(config.alert_cache_size)));
+        let alert_cache_expiry = config.alert_cache_expiry;
+        let watchtower_system_name = config.watchtower_system_name.to_string();
+        let allowed_alerting_start_time = SystemTime::now() + config.min_duration_from_start_to_err;
```

Comment:
Sometimes we use `Instant::now()` and other times we use `SystemTime::now()`. Any difference?